### PR TITLE
[move-compiler] Added support for linter warning suppression

### PR DESCRIPTION
**Pull request by awelc**
## Description 

This implements the ability to add suppression of external warnings, particularly warnings coming from Sui linters.

## Test Plan 

Must pass all existing tests plus updated linter tests.

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] protocol change
- [x] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### Release notes

Developers can now selectively suppress linter warnings. Linter warnings now also feature an additional message describing how they can be suppressed.

### CODE REVIEW

File: external-crates/move/move-compiler/src/shared/mod.rs [hunk 1]
```diff
@@ -200,13 +237,153 @@ impl CompilationEnv {
         default_config: Option<PackageConfig>,
     ) -> Self {
         visitors.extend([sui_mode::id_leak::IDLeakVerifier.into()]);
+        let filter_attr_name =
+            E::AttributeName_::Known(known_attributes::KnownAttribute::Diagnostic(
+                known_attributes::DiagnosticAttribute::Allow,
+            ));
+        let filter_attributes = BTreeSet::from([filter_attr_name]);
+        let known_filters = BTreeMap::from([
```
Comment (tnowacki) [inline external-crates/move/move-compiler/src/shared/mod.rs:245]: Could we get the macro (or a macro) back for declaring these? I just find it a hard to parse visually as is.

File: external-crates/move/move-compiler/src/shared/mod.rs [hunk 2]
```diff
@@ -200,13 +237,153 @@ impl CompilationEnv {
...
+            (
+                KnownFilterKey::new(FILTER_ALL.to_string(), filter_attr_name.clone()),
+                WarningFilter::All,
+            ),
+            (
+                KnownFilterKey::new(FILTER_UNUSED.to_string(), filter_attr_name.clone()),
```
Comment (tnowacki) [inline external-crates/move/move-compiler/src/shared/mod.rs:251]: AttributeName_ implements `Copy`, so you should be able to remove these clones

File: external-crates/move/move-compiler/src/shared/mod.rs [hunk 3]
```diff
@@ -179,6 +195,21 @@ pub struct IndexedPackagePath {
...
 pub type AttributeDeriver = dyn Fn(&mut CompilationEnv, &mut ModuleDefinition);
 
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct KnownFilterKey {
+    name: String,
+    attribute_name: E::AttributeName_,
+}
+
+impl KnownFilterKey {
+    pub fn new(name: String, attribute_name: E::AttributeName_) -> Self {
```
Comment (tnowacki) [inline external-crates/move/move-compiler/src/shared/mod.rs:205]:
```suggestion
    pub fn new(name: impl Into<String>, attribute_name: E::AttributeName_) -> Self {
```
or
```suggestion
    pub fn new(name: impl ToString, attribute_name: E::AttributeName_) -> Self {
```
would help clean up the call sites a bit

File: external-crates/move/move-compiler/src/diagnostics/mod.rs [hunk 4]
```diff
@@ -416,24 +419,29 @@ impl WarningFilters {
...
-                    category.insert(cat);
+                    category.insert(cat, n);
                     // remove any codes now covered by this category
-                    codes.retain(|(codes_cat, _)| codes_cat != &cat);
+                    codes.retain(
+                        |DiagnosticsID {
+                             category: codes_cat,
+                             code: _,
+                         },
+                         _| codes_cat != &cat,
```
Comment (tnowacki) [inline external-crates/move/move-compiler/src/diagnostics/mod.rs:438]:
```suggestion
                        |diag_id, _| diag_id.cateogry != &cat,
```
Love the pattern matching :) But this might be a bit more readable

File: external-crates/move/move-compiler/src/diagnostics/mod.rs [hunk 5]
```diff
@@ -59,11 +59,11 @@ pub struct Diagnostics {
 pub enum WarningFilters {
```
Comment (tnowacki) [inline external-crates/move/move-compiler/src/diagnostics/mod.rs:59]: I think the main thing missing here is we have to domain separate the `WarningFilters`. As implemented today, the warning category/codes will overlap between external tools and the ones defined by the compiler.

File: external-crates/move/move-compiler/src/shared/mod.rs [hunk 6]
```diff
@@ -179,6 +195,21 @@ pub struct IndexedPackagePath {
 
 pub type AttributeDeriver = dyn Fn(&mut CompilationEnv, &mut ModuleDefinition);
 
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct KnownFilterKey {
+    name: String,
```
Comment (tnowacki) [inline external-crates/move/move-compiler/src/shared/mod.rs:200]: Should we use `Symbol` here?

File: external-crates/move/move-compiler/src/diagnostics/codes.rs [hunk 7]
```diff
@@ -14,11 +14,16 @@ pub enum Severity {
...
+}
+
 #[derive(PartialEq, Eq, Clone, Debug, Hash)]
 pub struct DiagnosticInfo {
     severity: Severity,
-    category: u8,
-    code: u8,
+    id: DiagnosticsID,
     message: &'static str,
     external_prefix: Option<&'static str>,
```
Comment (tnowacki) [inline external-crates/move/move-compiler/src/diagnostics/codes.rs:28]: We could maybe fold in `external_prefix: Option<&'static str>,` into `DiagnosticsID` as a solution to the domain separation problem

File: external-crates/move/move-compiler/src/shared/mod.rs (hunk 3 above)
Comment (awelc) [reply external-crates/move/move-compiler/src/shared/mod.rs:205]: Cool trick! But in this case using `&str` should suffice as well to clean up the call sites

File: external-crates/move/move-compiler/src/shared/mod.rs (hunk 6 above)
Comment (awelc) [reply external-crates/move/move-compiler/src/shared/mod.rs:200]: We certainly can!

File: external-crates/move/move-compiler/src/diagnostics/mod.rs (hunk 5 above)
Comment (awelc) [reply external-crates/move/move-compiler/src/diagnostics/mod.rs:61]: Indeed. A new version fixes this.

File: external-crates/move/move-compiler/src/diagnostics/codes.rs (hunk 7 above)
Comment (awelc) [reply external-crates/move/move-compiler/src/diagnostics/codes.rs]: It's a good start but we need to domain-separate the `All` filter and the `Category` filter as well. I followed a similar path to add `external_prefix` to both and added extra tests for checking this.

File: external-crates/move/move-compiler/src/shared/mod.rs (hunk 1 above)
Comment (awelc) [reply external-crates/move/move-compiler/src/shared/mod.rs:246]: I tried...

File: crates/sui-move-build/tests/linter/share_owned.exp [hunk 8]
```diff
@@ -9,6 +9,8 @@ warning[Lint W01001]: possible owned object share
    │         │                             │
    │         │                             Creating a fresh object and sharing it within the same function will ensure this does not abort.
    │         Potential abort from a (potentially) owned object created by a different transaction.
+   │
+   = This warning can be suppressed with '#[allow(share_owned)]' applied to the 'module' or module member ('const', 'fun', or 'struct')
```
Comment (tnowacki) [inline crates/sui-move-build/tests/linter/share_owned.exp:13]: I think we need to fix this message? 
```suggestion
   = This warning can be suppressed with '#[lint_allow(share_owned)]' applied to the 'module' or module member ('const', 'fun', or 'struct')
```

File: external-crates/move/move-compiler/src/diagnostics/codes.rs [hunk 9]
```diff
@@ -14,13 +14,26 @@ pub enum Severity {
     Bug = 3,
 }
 
+/// Identifies a warning category. Includes an option external prefix to distinguish warnings from
+/// different sources.
+#[derive(PartialEq, Eq, Clone, Copy, Debug, Hash, PartialOrd, Ord)]
+pub struct CategoryID {
+    category: u8,
+    external_prefix: Option<&'static str>,
```
Comment (tnowacki) [inline external-crates/move/move-compiler/src/diagnostics/codes.rs:22]: Nit,  maybe a type alias for this, then we wouldn't need some of the comments below (maybe)

File: external-crates/move/move-compiler/src/shared/mod.rs (hunk 1 above)
Comment (tnowacki) [reply external-crates/move/move-compiler/src/shared/mod.rs:258]: This looks a lot better, thanks!

File: external-crates/move/move-compiler/src/shared/mod.rs [hunk 10]
```diff
@@ -219,16 +347,13 @@ impl CompilationEnv {
...
-                    diag.info().code(),
-                );
-                if let Some(filter_name) = possible_filter.to_str() {
+            if diag.info().severity() == Severity::Warning {
+                if let Some(filter_name) = self.known_filter_names.get(&diag.info().id()) {
                     let help = format!(
                         "This warning can be suppressed with '#[{}({})]' \
                         applied to the 'module' or module member ('const', 'fun', or 'struct')",
-                        WARNING_FILTER_ATTR, filter_name
+                        WARNING_FILTER_ATTR,
```
Comment (tnowacki) [inline external-crates/move/move-compiler/src/shared/mod.rs:355]: Just noting again that we need the prefix here now

File: crates/sui-move-build/tests/linter/share_owned.exp (hunk 8 above)
Comment (awelc) [reply crates/sui-move-build/tests/linter/share_owned.exp:13]: Good point!

File: external-crates/move/move-compiler/src/shared/mod.rs (hunk 10 above)
Comment (awelc) [reply external-crates/move/move-compiler/src/shared/mod.rs:355]: Noted!

**Review by tnowacki — APPROVED** [review]
LGTM! Thanks!
